computer-use-demo/computer_use_demo/loop.py

```diff
@@ -74,71 +83,120 @@ async def sampling_loop(
+
+            api_response_callback(cast(APIResponse[BetaMessage], raw_response))
+
+            response = raw_response.parse()
+            debug_print(f"[DEBUG] Parsed API response: {response}")
+
+            messages.append(
+                {
+                    "role": "assistant",
+                    "content": cast(list[BetaContentBlockParam], response.content),
+                }
+            )
+            debug_print(f"[DEBUG] Updated messages: {messages}")
+
+            tool_result_content: list[BetaToolResultBlockParam] = []
+            for content_block in cast(list[BetaContentBlock], response.content):
+                debug_print(f"[DEBUG] Processing content_block: {content_block}")
+                output_callback(content_block)
+                if content_block.type == "tool_use":
+                    debug_print(f"[DEBUG] Tool use detected: {content_block.name} with input {content_block.input}")
+                    result = await tool_collection.run(
+                        name=content_block.name,
+                        tool_input=cast(dict[str, Any], content_block.input),
+                    )
+                    debug_print(f"[DEBUG] Tool result: {result}")
+                    tool_result_content.append(
+                        _make_api_tool_result(result, content_block.id)
+                    )
+                    tool_output_callback(result, content_block.id)
+
+            if not tool_result_content:
+                debug_print("[DEBUG] No tool results to process. Exiting sampling loop.")
+                return messages
+
+            debug_print(f"[DEBUG] Appending tool results to messages: {tool_result_content}")
+            messages.append({"content": tool_result_content, "role": "user"})
+            retry_count = 0  # Reset retry count on success
+
+        except Exception as e:
+            error_message = str(e).lower()
+            debug_print(f"[DEBUG] Exception caught in sampling_loop: {e}")
+
+            if "rate_limit_error" in error_message:
+                if "daily rate limit" in error_message:
+                    debug_print("[DEBUG] Daily rate limit exceeded. Exiting program.")
+                    sys.exit(1)
+                retry_count += 1
+                delay = min(2**retry_count, 60)  # Exponential backoff with 60s cap
+
+                debug_print(
+                    f"[DEBUG] Rate limit hit. Retry attempt: {retry_count}, "
+                    f"Delaying for {delay} seconds."
                 )
-                tool_output_callback(result, content_block.id)
 
-        if not tool_result_content:
-            return messages
+                if retry_count > MAX_RETRIES:
+                    debug_print("[DEBUG] Maximum retry attempts exceeded. Exiting program.")
+                    sys.exit(1)
 
-        messages.append({"content": tool_result_content, "role": "user"})
+                await asyncio.sleep(delay)  # Introduce delay before retry
```

Comment:
note that it is preferable to read the `retry-after` header from the error (i.e. `RateLimitError` type in the SDK) if available.